File: utils/keys.ts
// Physical key definitions used across the Rules and Keymap tabs.
//
// Layout uses a flat grid (NOT a realistic keyboard layout) split into
// "left hand" and "right hand" halves. Keys are identified by stable codes
// (roughly following KeyboardEvent.code naming for letters/digits, plus
// explicit names for modifiers / function keys).

export interface KeyDef {
  // Stable code used as map key in LayerKeymap.keys.
  code: string
  // Label shown on the cap (English).
  label: string
}

export type KeyLabelMode = 'label' | 'code' | 'numeric'

const KEY_NUMERIC_CODES: Record<string, number> = {
  Escape: 1,
  Digit1: 2,
  Digit2: 3,
  Digit3: 4,
  Digit4: 5,
  Digit5: 6,
  Digit6: 7,
  Digit7: 8,
  Digit8: 9,
  Digit9: 10,
  Digit0: 11,
  Minus: 12,
  Equal: 13,
  Backspace: 14,
  Tab: 15,
  KeyQ: 16,
  KeyW: 17,
  KeyE: 18,
  KeyR: 19,
  KeyT: 20,
  KeyY: 21,
  KeyU: 22,
  KeyI: 23,
  KeyO: 24,
  KeyP: 25,
  BracketLeft: 26,
  BracketRight: 27,
  Enter: 28,
  ControlLeft: 29,
  KeyA: 30,
  KeyS: 31,
  KeyD: 32,
  KeyF: 33,
  KeyG: 34,
  KeyH: 35,
  KeyJ: 36,
  KeyK: 37,
  KeyL: 38,
  Semicolon: 39,
  Quote: 40,
  Backquote: 41,
  ShiftLeft: 42,
  Backslash: 43,
  KeyZ: 44,
  KeyX: 45,
  KeyC: 46,
  KeyV: 47,
  KeyB: 48,
  KeyN: 49,
  KeyM: 50,
  Comma: 51,
  Period: 52,
  Slash: 53,
  ShiftRight: 54,
  AltLeft: 56,
  Space: 57,
  CapsLock: 58,
  F1: 59,
  F2: 60,
  F3: 61,
  F4: 62,
  F5: 63,
  F6: 64,
  F7: 65,
  F8: 66,
  F9: 67,
  F10: 68,
  F11: 87,
  F12: 88,
  ControlRight: 97,
  AltRight: 100,
  MetaLeft: 125,
  MetaRight: 126,
  ContextMenu: 127,
  PrintScreen: 99,
}

function row(codes: Array<[string, string]>): KeyDef[] {
  return codes.map(([code, label]) => ({ code, label }))
}

export const LEFT_HAND_ROWS: KeyDef[][] = [
  row([
    ['Escape', 'Esc'],
    ['F1', 'F1'],
    ['F2', 'F2'],
    ['F3', 'F3'],
    ['F4', 'F4'],
    ['F5', 'F5'],
  ]),
  row([
    ['Backquote', '`'],
    ['Digit1', '1'],
    ['Digit2', '2'],
    ['Digit3', '3'],
    ['Digit4', '4'],
    ['Digit5', '5'],
  ]),
  row([
    ['Tab', 'Tab'],
    ['KeyQ', 'Q'],
    ['KeyW', 'W'],
    ['KeyE', 'E'],
    ['KeyR', 'R'],
    ['KeyT', 'T'],
  ]),
  row([
    ['CapsLock', 'Caps'],
    ['KeyA', 'A'],
    ['KeyS', 'S'],
    ['KeyD', 'D'],
    ['KeyF', 'F'],
    ['KeyG', 'G'],
  ]),
  row([
    ['ShiftLeft', 'Shift'],
    ['KeyZ', 'Z'],
    ['KeyX', 'X'],
    ['KeyC', 'C'],
    ['KeyV', 'V'],
    ['KeyB', 'B'],
  ]),
  row([
    ['ControlLeft', 'Ctrl'],
    ['MetaLeft', 'Meta'],
    ['AltLeft', 'Alt'],
    ['Space', 'Space'],
  ]),
]

export const RIGHT_HAND_ROWS: KeyDef[][] = [
  row([
    ['F6', 'F6'],
    ['F7', 'F7'],
    ['F8', 'F8'],
    ['F9', 'F9'],
    ['F10', 'F10'],
    ['F11', 'F11'],
    ['F12', 'F12'],
    ['PrintScreen', 'PrtSc'],
  ]),
  row([
    ['Digit6', '6'],
    ['Digit7', '7'],
    ['Digit8', '8'],
    ['Digit9', '9'],
    ['Digit0', '0'],
    ['Minus', '-'],
    ['Equal', '='],
    ['Backspace', 'Bksp'],
  ]),
  row([
    ['KeyY', 'Y'],
    ['KeyU', 'U'],
    ['KeyI', 'I'],
    ['KeyO', 'O'],
    ['KeyP', 'P'],
    ['BracketLeft', '['],
    ['BracketRight', ']'],
    ['Backslash', '\\'],
  ]),
  row([
    ['KeyH', 'H'],
    ['KeyJ', 'J'],
    ['KeyK', 'K'],
    ['KeyL', 'L'],
    ['Semicolon', ';'],
    ['Quote', "'"],
    ['Enter', 'Enter'],
  ]),
  row([
    ['KeyN', 'N'],
    ['KeyM', 'M'],
    ['Comma', ','],
    ['Period', '.'],
    ['Slash', '/'],
    ['ShiftRight', 'Shift'],
  ]),
  row([
    ['AltRight', 'Alt'],
    ['MetaRight', 'Meta'],
    ['ContextMenu', 'Menu'],
    ['ControlRight', 'Ctrl'],
  ]),
]

export const ALL_KEYS: KeyDef[] = [
  ...LEFT_HAND_ROWS.flat(),
  ...RIGHT_HAND_ROWS.flat(),
]

export function keyLabel(code: string, mode: KeyLabelMode = 'label'): string {
  if (mode === 'code') return code
  if (mode === 'numeric') return String(KEY_NUMERIC_CODES[code] ?? code)
  return ALL_KEYS.find((k) => k.code === code)?.label ?? code
}

export function randomId(): string {
  return Math.random().toString(36).slice(2, 10)
}

File: src-tauri/src/mapper/keys.rs
// Mapping between frontend key codes (KeyboardEvent.code-ish, see utils/keys.ts)
// and Linux evdev `Key` values.

#![cfg(target_os = "linux")]

use evdev::Key;

pub fn code_to_key(code: &str) -> Option<Key> {
    Some(match code {
        // Letters
        "KeyA" => Key::KEY_A, "KeyB" => Key::KEY_B, "KeyC" => Key::KEY_C,
        "KeyD" => Key::KEY_D, "KeyE" => Key::KEY_E, "KeyF" => Key::KEY_F,
        "KeyG" => Key::KEY_G, "KeyH" => Key::KEY_H, "KeyI" => Key::KEY_I,
        "KeyJ" => Key::KEY_J, "KeyK" => Key::KEY_K, "KeyL" => Key::KEY_L,
        "KeyM" => Key::KEY_M, "KeyN" => Key::KEY_N, "KeyO" => Key::KEY_O,
        "KeyP" => Key::KEY_P, "KeyQ" => Key::KEY_Q, "KeyR" => Key::KEY_R,
        "KeyS" => Key::KEY_S, "KeyT" => Key::KEY_T, "KeyU" => Key::KEY_U,
        "KeyV" => Key::KEY_V, "KeyW" => Key::KEY_W, "KeyX" => Key::KEY_X,
        "KeyY" => Key::KEY_Y, "KeyZ" => Key::KEY_Z,

        // Digits
        "Digit0" => Key::KEY_0, "Digit1" => Key::KEY_1, "Digit2" => Key::KEY_2,
        "Digit3" => Key::KEY_3, "Digit4" => Key::KEY_4, "Digit5" => Key::KEY_5,
        "Digit6" => Key::KEY_6, "Digit7" => Key::KEY_7, "Digit8" => Key::KEY_8,
        "Digit9" => Key::KEY_9,

        // F-row
        "F1" => Key::KEY_F1, "F2" => Key::KEY_F2, "F3" => Key::KEY_F3,
        "F4" => Key::KEY_F4, "F5" => Key::KEY_F5, "F6" => Key::KEY_F6,
        "F7" => Key::KEY_F7, "F8" => Key::KEY_F8, "F9" => Key::KEY_F9,
        "F10" => Key::KEY_F10, "F11" => Key::KEY_F11, "F12" => Key::KEY_F12,

        // Navigation / editing
        "Escape" | "Esc" => Key::KEY_ESC,
        "Tab" => Key::KEY_TAB,
        "CapsLock" => Key::KEY_CAPSLOCK,
        "Enter" | "Return" => Key::KEY_ENTER,
        "Backspace" => Key::KEY_BACKSPACE,
        "Space" => Key::KEY_SPACE,
        "Delete" | "Del" => Key::KEY_DELETE,
        "Insert" | "Ins" => Key::KEY_INSERT,
        "Home" => Key::KEY_HOME,
        "End" => Key::KEY_END,
        "PageUp" | "PgUp" => Key::KEY_PAGEUP,
        "PageDown" | "PgDn" => Key::KEY_PAGEDOWN,
        "ArrowLeft" | "Left" => Key::KEY_LEFT,
        "ArrowRight" | "Right" => Key::KEY_RIGHT,
        "ArrowUp" | "Up" => Key::KEY_UP,
        "ArrowDown" | "Down" => Key::KEY_DOWN,

        // Punctuation
        "Backquote" => Key::KEY_GRAVE,
        "Minus" => Key::KEY_MINUS,
        "Equal" => Key::KEY_EQUAL,
        "BracketLeft" => Key::KEY_LEFTBRACE,
        "BracketRight" => Key::KEY_RIGHTBRACE,
        "Backslash" => Key::KEY_BACKSLASH,
        "Semicolon" => Key::KEY_SEMICOLON,
        "Quote" => Key::KEY_APOSTROPHE,
        "Comma" => Key::KEY_COMMA,
        "Period" => Key::KEY_DOT,
        "Slash" => Key::KEY_SLASH,

        // Modifiers
        "ShiftLeft" | "Shift" => Key::KEY_LEFTSHIFT,
        "ShiftRight" => Key::KEY_RIGHTSHIFT,
        "ControlLeft" | "Control" | "Ctrl" => Key::KEY_LEFTCTRL,
        "ControlRight" => Key::KEY_RIGHTCTRL,
        "AltLeft" | "Alt" => Key::KEY_LEFTALT,
        "AltRight" => Key::KEY_RIGHTALT,
        "MetaLeft" | "Meta" | "Super" | "Win" => Key::KEY_LEFTMETA,
        "MetaRight" => Key::KEY_RIGHTMETA,
        "ContextMenu" | "Menu" => Key::KEY_COMPOSE,

        // Media / browser
        "BrowserBack" => Key::KEY_BACK,
        "BrowserForward" => Key::KEY_FORWARD,
        "BrowserRefresh" => Key::KEY_REFRESH,
        "BrowserHome" => Key::KEY_HOMEPAGE,
        "VolumeUp" => Key::KEY_VOLUMEUP,
        "VolumeDown" => Key::KEY_VOLUMEDOWN,
        "VolumeMute" | "Mute" => Key::KEY_MUTE,
        "MediaPlayPause" | "PlayPause" => Key::KEY_PLAYPAUSE,
        "MediaNext" | "Next" => Key::KEY_NEXTSONG,
        "MediaPrev" | "Prev" | "Previous" => Key::KEY_PREVIOUSSONG,

        _ => return None,
    })
}

// Single-character shortcuts (e.g. the symbols in default-layers.yaml).
// Returns (needs_shift, Key).
pub fn char_to_key(ch: char) -> Option<(bool, Key)> {
    Some(match ch {
        // Unshifted
        '`' => (false, Key::KEY_GRAVE),
        '-' => (false, Key::KEY_MINUS),
        '=' => (false, Key::KEY_EQUAL),
        '[' => (false, Key::KEY_LEFTBRACE),
        ']' => (false, Key::KEY_RIGHTBRACE),
        '\\' => (false, Key::KEY_BACKSLASH),
        ';' => (false, Key::KEY_SEMICOLON),
        '\'' => (false, Key::KEY_APOSTROPHE),
        ',' => (false, Key::KEY_COMMA),
        '.' => (false, Key::KEY_DOT),
        '/' => (false, Key::KEY_SLASH),
        ' ' => (false, Key::KEY_SPACE),

        // Shifted
        '~' => (true, Key::KEY_GRAVE),
        '!' => (true, Key::KEY_1),
        '@' => (true, Key::KEY_2),
        '#' => (true, Key::KEY_3),
        '$' => (true, Key::KEY_4),
        '%' => (true, Key::KEY_5),
        '^' => (true, Key::KEY_6),
        '&' => (true, Key::KEY_7),
        '*' => (true, Key::KEY_8),
        '(' => (true, Key::KEY_9),
        ')' => (true, Key::KEY_0),
        '_' => (true, Key::KEY_MINUS),
        '+' => (true, Key::KEY_EQUAL),
        '{' => (true, Key::KEY_LEFTBRACE),
        '}' => (true, Key::KEY_RIGHTBRACE),
        '|' => (true, Key::KEY_BACKSLASH),
        ':' => (true, Key::KEY_SEMICOLON),
        '"' => (true, Key::KEY_APOSTROPHE),
        '<' => (true, Key::KEY_COMMA),
        '>' => (true, Key::KEY_DOT),
        '?' => (true, Key::KEY_SLASH),

        c if c.is_ascii_digit() => (false, match c {
            '0' => Key::KEY_0, '1' => Key::KEY_1, '2' => Key::KEY_2,
            '3' => Key::KEY_3, '4' => Key::KEY_4, '5' => Key::KEY_5,
            '6' => Key::KEY_6, '7' => Key::KEY_7, '8' => Key::KEY_8,
            '9' => Key::KEY_9, _ => unreachable!(),
        }),
        c if c.is_ascii_lowercase() => {
            // Map a..z to KEY_A..KEY_Z via numeric code offset.
            let base = Key::KEY_A.code();
            let key = Key::new(base + (c as u16 - b'a' as u16));
            (false, key)
        }
        c if c.is_ascii_uppercase() => {
            let base = Key::KEY_A.code();
            let key = Key::new(base + (c as u16 - b'A' as u16));
            (true, key)
        }

        _ => return None,
    })
}
